Masonry use calculated height instead of estimated height

stwrt opened the pull request:
#### 3c270c15b7888c72d8d10e29bc43e40dba71e1d2
<pre>
Masonry use calculated height instead of estimated height
<a href="https://bugs.webkit.org/show_bug.cgi?id=250720">https://bugs.webkit.org/show_bug.cgi?id=250720</a>

Reviewed by Matt Woodrow.

The masonry calculations were being done before we had a done a proper layout.
With some refactoring the masonry layout was pushed back in the chain. This allows
the children content to have calculated the widths and heights. Now, we can just
use the real widths and heights instead of relying on the estimated ones.

* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::gridAreaBreadthForChild const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008.html: Added.

Canonical link: <a href="https://commits.webkit.org/259308@main">https://commits.webkit.org/259308@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95fc85e819d9be163efff59b2552ec233e983b9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104579 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113857 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4583 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96917 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112809 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110344 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94439 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108056 "webkitpy-tests (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93253 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26051 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7013 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27409 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92469 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4793 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7131 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30052 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103416 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13171 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46965 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101155 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6417 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8920 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -1588,6 +1588,8 @@ imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-r
 imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-003.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-columns-item-placement-auto-flow-next-001.html [ Skip ]
+imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-006.html [ Crash ]
+imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-007.html [ Crash ]
```
This seems unexpected

stwrt replied:
This was only reproducing in WebKit Test Runner when not attached to a debugger. Could not reproduce this at all using Minibrowser.

The error message mentioned a dirty renderer. I spent some time looking into it, and have not made a ton of progress so far in fixing this error.

nt1m replied:
Dirty renderer means you're entering layout with that renderer being dirty (e.g. has a "needs layout" flag that hasn't been cleared). Something must have added the "needs layout" flag, without ever clearing it, which is something that needs to be fixed.

stwrt replied:
Setting `MarkContainingBlockChain` was causing the issues. I think this is still needed, but this is ultimately a different issue, and I can handle it next patch.

mattwoodrow commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -168,11 +168,12 @@ void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition()
+    
     // FIXME(249230): Try to cache masonry layout sizes
-    child.setChildNeedsLayout(MarkOnlyThis);
+    child.setChildNeedsLayout(MarkContainingBlockChain);
```
I don't think you need this change, and I'm 95% sure it's the cause of your crash.

stwrt replied:
We needed this for some inner grid fixes, but reverted for now.

I will tackle this in a follow up patch.

Loirooriol reviewed (commented):
Why no test? And no explanation for a lack of test in the commit message.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008-ref.html:
```diff
@@ -0,0 +1,28 @@
+  <meta charset="utf-8">
+  <title>CSS Grid Test: Masonry item placement w/ Images</title>
+  <link rel="author" title="Brandon Stewart" href="mailto:brandonstewart@apple.com">
+  <link rel="help" href="https://drafts.csswg.org/css-grid-2">
```
You'll also need a copy of this file named `-expected.html` in order for WebKit to actually run this test.

Also please export the tests to the WPT repo (`export-w3c-test-changes`)

stwrt replied:
Added the file

Loirooriol replied:
Still missing the -expected.html and WPT export.
Also, the commit message doesn't lists the modified test files.

stwrt replied:
Not sure what happened to the expected when I update this commit. It should be back now.

mattwoodrow reviewed (approved):
Please update the changed file list to include the test files.

stwrt committed:
Masonry use calculated height instead of estimated height
https://bugs.webkit.org/show_bug.cgi?id=250720

Reviewed by Matt Woodrow.

The masonry calculations were being done before we had a done a proper layout.
With some refactoring the masonry layout was pushed back in the chain. This allows
the children content to have calculated the widths and heights. Now, we can just
use the real widths and heights instead of relying on the estimated ones.

* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::gridAreaBreadthForChild const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/item-placement/masonry-item-placement-008.html: Added.

Canonical link: https://commits.webkit.org/259308@main
